### ISSUE TITLE
Combo of HTML Editor and StripHTML Causes Missing Whitespace

### ISSUE THREAD

**Issue by JRichmond27** [SparkDevNetwork/Rock#4633]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Did you perform a search at https://github.com/issues?q=is%3Aissue+user%3ASparkDevNetwork+-repo%3ARock to see if your bug or enhancement is already reported?
    * Can you reproduce the problem on a fresh install or the [demo site](http://rock.rocksolidchurchdemo.com/)?
    * Did you include your Rock version number and [client culture](https://github.com/SparkDevNetwork/Rock/wiki/Environment-and-Diagnostics-Information) setting?

### Description

Not sure if this would be considered a bug, but it at least seems like an unwanted side-effect... 

The HTML editor in Rock likes to create HTML elements without any whitespace between the tags. I'm assuming this is either to reduce file size or to prevent page formatting issues that can be caused by extra whitespace. Unfortunately I recently noticed that this creates an unwanted side effect when the `StripHtml` Lava filter is used. Text from adjacent HTML elements end up concatenated together with no space in between.

### Steps to Reproduce

**Quick Test:**
1. Add an HTML Content block to a page
2. Add this code to the content block
```
{%- assign test1 = '<p>Test line one.</p><p>Test line two.</p>' -%}
{%- capture test2 %}<p>Test line one.</p>
<p>Test line two.</p>{% endcapture -%}
[{{ test1 | StripHtml }}]
<br>
[{{ test2 | StripHtml }}]
```
3. Save and load the page
4. Notice that the first test does not contain a space between "one." and "Test", but the second test does.

**Complete Test:**
1. Create a Content Channel Item with a Content field that includes more than one HTML element without any trailing whitespace inside the elements.
2. Add a Content Channel Item View block to a page
3. In the Lava Template for the block, add Lava to display the Item Content with the `StripHTML` filter applied
4. Load the page with the Content Channel Item that you created
5. Notice that there is no space between the text from the separate HTML elements

**Expected behavior:**

Even when there is no space between HTML elements, I would expect `StripHTML` to include a space AND/OR I would expect the HTML editor to include whitespace between elements.

**Actual behavior:**

The HTML editor concatenates all HTML elements into one long string with no spaces between the elements and using `StripHtml` on that content results in text that is missing spaces where there should be spaces.

### Versions

* **Rock Version:** Rock McKinley 11.4 (1.11.4.3)
* **Client Culture Setting:** en-US

**Comment by cwomack** [comment]
Hey Jeff (@JRichmond27),

It appears that it's the browser is the culprit for the above referenced changes to the output HTML within Rock.  The browser is specified to turn a new line to a white space as line breaks are also white space characters.  Feel free to reply back if you find anything else that points to Rock generating the issue, but I'll close this out since it doesn't seem to be a bug!

**Comment by cwomack** [comment]
Here's a couple resources that can help explain it further:  

- [MDN Article](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Whitespace)
- [Stack Overflow](https://stackoverflow.com/a/588378)
- [w3.org](https://www.w3.org/TR/REC-html40/struct/text.html#h-9.1)

**Comment by JRichmond27** [comment]
Hey @cwomack, thanks for your response, 

It sounds like I may not have described the issue very well. I understand how whitespace works in browsers, but this is more related to the actual HTML code stored in the database and/or how it is handled when used as input for the `StripHTML` filter, not how the browser interprets or renders the HTML.

If you take the HTML generated by the editor and view it in a browser, it will render perfectly fine. Paragraph breaks will be rendered with carriage return/line feed whitespace between them exactly as expected.

The problem is a combo of the actual HTML string that is generated by the editor and how the `StripHTML` filter handles HTML that is generated in that way. 

The editor generates HTML with absolutely no whitespace characters between closing and opening tags, which is fine, but then `StripHTML` removes all characters related to HTML elements without taking into account where it might be a good idea to include a space character. 

So for example, if I type a line of text into the editor, then hit the enter key, and then type a second line, this is the actual HTML that is generated in the background and ultimately saved to the database:
```
<p>My first line of text.</p><p>My second line of text.</p>
```
Even though the long lines of unformatted run-on HTML throw my perfectionist into distress, it is completely valid HTML and will render perfectly when viewed in a browser. However, when the same HTML is run through the `StripHTML` filter, this is the actual text output:
```
My first line of text.My second line of text.
```
Note the lack of a space between "text." and "My". I don't necessarily think this is technically a bug, but it's definitely unexpected/undesired behavior that seems like should be avoided.

I mention both the editor and the filter because it seems like the solution could come from either of those places...
1. The editor could be updated so that the HTML for certain adjacent elements (at least for block level elements) is generated with either a space or a newline character between the closing tag of one and the opening tag of the next.
2. The `StripHTML` filter could be updated to automatically insert a space in the places where it removes some or all HTML elements

The results of either solution would be pretty close to the same. The HTML viewed in a browser would still render just like it does now (depending on the CSS), but the new output of the filter would be much more expected:
```
My first line of text. My second line of text.
```

My personal preference would probably be the first option since it would produce more readable HTML (for us dev nerds) and depending on how the filter is written, I'm guessing the second option might cause a little extra processing on every page load.

I hope that explains things better. Please let me know if you have more questions or need any more clarification. Thanks!

**Comment by nairdo** [comment]
Hi Jeff @JRichmond27 Let me see if I can clarify this a bit more using your examples and for the sake of anyone else reading.

For the first test string (`<p>Test line one.</p><p>Test line two.</p>`) the `StripHtml` filter would/should turn that into `Test line one.Test line two.`  That is the correct behavior.  However, I understand the idea your proposing and will review that internally.

For the second test string, the `StripHtml` filter is again only removing the `<p>` and `</p>` tags -- so the line-break in your text will remain.  You can this by looking at the 'view-source':

![image](https://user-images.githubusercontent.com/383807/112396354-33298a00-8cbd-11eb-9e98-1c4a2a442655.png)

...and I only point that out because, at first glance, it seems odd that it would render with a space.

**Comment by JRichmond27** [comment]
Yes, I included the second test as more of a control test. `StripHTML` works as I would expect with that one. The first case is the one that doesn't work as expected (in my opinion).

When looking at it from a non-techie point of view, if I was a typical staff member and I created a blog post with multiple paragraphs, and then the content of that blog post was displayed somewhere on the website using `StripHTML`, I would not expect to see the end of one paragraph butted right up against the next paragraph without any space in between.

I've been thinking about implementing a workaround like your example, except that there are a lot more elements to consider in addition to paragraphs. I may end up trying some kind of regex variation of that solution.

Thanks again for the responses!
